In [1]:
from pathlib import Path
import pandas as pd
import numpy as np

In [3]:
flightlist = pd.concat(pd.read_csv(file, 
                                   parse_dates=['firstseen','lastseen','day']) for file in Path('C:/Users/ldmag/Desktop/Desktop Icons/Datasets/AirlineInformation').glob('flightlist_*.csv'))


<ipython-input-3-ef3062578422>:1: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.
  flightlist = pd.concat(pd.read_csv(file,


In [4]:
flightlist

callsign  number  icao24 registration typecode origin destination  \
0          HVN19     NaN  888152          NaN      NaN   YMML        LFPG   
1         CCA839     NaN  780ad1          NaN      NaN   YMML        LEBL   
2         CES219     NaN  780b7e       B-5936     A332   YSSY        EDDF   
3         AEA040     NaN  34444e       EC-LVL     A332   LEMD        LEMD   
4         CXA825     NaN  780d75       B-2760     B788   YSSY        LFPG   
...          ...     ...     ...          ...      ...    ...         ...   
1894746   N114RG     NaN  a03c8d       N114RG      NaN   KSQL        CL77   
1894747    ZKOPM     NaN  c82200          NaN      NaN    NaN        NZAR   
1894748  CSC8985     NaN  780ce9       B-1882     A320    NaN         NaN   
1894749  CDG8413     NaN  7bb130       B-5111     B738    NaN         NaN   
1894750  SKW4228  OO4228  a602cd       N487CA     CRJ2   KSLC         NaN   

                        firstseen                  lastseen  \
0       2018-12-31 00:43:16+00:00 2019-01-01 04:56:29+00:00   
1       2018-12-31 00:53:08+00:00 2019-01-01 06:01:42+00:00   
2       2018-12-31 01:05:29+00:00 2019-01-01 04:09:29+00:00   
3       2018-12-31 01:07:21+00:00 2019-01-01 03:32:59+00:00   
4       2018-12-31 01:18:29+00:00 2019-01-01 04:32:28+00:00   
...                           ...                       ...   
1894746 2020-12-31 23:44:46+00:00 2020-12-31 23:57:53+00:00   
1894747 2020-12-31 23:45:54+00:00 2020-12-31 23:58:58+00:00   
1894748 2020-12-31 23:46:40+00:00 2020-12-31 23:59:19+00:00   
1894749 2020-12-31 23:46:51+00:00 2020-12-31 23:58:59+00:00   
1894750 2020-12-31 23:46:53+00:00 2020-12-31 23:57:52+00:00   

                              day  latitude_1  longitude_1  altitude_1  \
0       2019-01-01 00:00:00+00:00  -37.659485   144.804421       304.8   
1       2019-01-01 00:00:00+00:00  -37.692123   144.841997       304.8   
2       2019-01-01 00:00:00+00:00  -33.930908   151.171987         0.0   
3       2019-01-01 00:00:00+00:00   40.534756    -3.575426       609.6   
4       2019-01-01 00:00:00+00:00  -33.954254   151.178041         0.0   
...                           ...         ...          ...         ...   
1894746 2020-12-31 00:00:00+00:00   37.512543  -122.250308      -304.8   
1894747 2020-12-31 00:00:00+00:00  -36.956726   175.387145       914.4   
1894748 2020-12-31 00:00:00+00:00   30.956863   105.965289      7315.2   
1894749 2020-12-31 00:00:00+00:00   24.899442   118.627625      4572.0   
1894750 2020-12-31 00:00:00+00:00   40.821533  -111.976379      1524.0   

         latitude_2  longitude_2  altitude_2  
0         48.995316     2.610802      -53.34  
1         41.036124     2.063557     1112.52  
2         50.045563     8.588923      -15.24  
3         40.475728    -3.538347      411.48  
4         48.996091     2.625805       22.86  
...             ...          ...         ...  
1894746   37.023321  -122.095244     1059.18  
1894747  -37.165329   174.898019      342.90  
1894748   29.713549   107.287095     7498.08  
1894749   26.105301   119.509226     7498.08  
1894750   40.595879  -111.068282     8572.50  

[53484007 rows x 16 columns]

In [5]:
from traffic.data import airports
hubs = [['KATL','KDFW','KORD','KLGA','KEWR']]

data = pd.concat((flightlist.query(f'origin == "{airport}"').groupby('day').agg(dict(callsign='count')).rename(columns=dict(callsign=airport)) for airport in sum(hubs,[])),axis=1)


In [6]:
data # some nans here, drop data before that

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


KATL  KDFW  KORD  KLGA  KEWR
day                                                     
2019-01-01 00:00:00+00:00    NaN   585   604   208   294
2019-01-02 00:00:00+00:00    NaN   747   665   276   479
2019-01-03 00:00:00+00:00    NaN   779   770   284   490
2019-01-04 00:00:00+00:00    NaN   793   758   278   430
2019-01-05 00:00:00+00:00    NaN   692   754   251   353
...                          ...   ...   ...   ...   ...
2020-12-27 00:00:00+00:00  887.0   812   677   141   325
2020-12-28 00:00:00+00:00  897.0   817   687   144   301
2020-12-29 00:00:00+00:00  923.0   862   684   147   337
2020-12-30 00:00:00+00:00  894.0   742   607   149   350
2020-12-31 00:00:00+00:00  792.0   788   637   128   322

[730 rows x 5 columns]

In [7]:
data.isnull().sum() * 100 / len(data)

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


KATL    13.835616
KDFW     0.000000
KORD     0.000000
KLGA     0.000000
KEWR     0.000000
dtype: float64

KATL has no data collected for a few months in 2019 - plot may contain this information so it shouldn't be taken at face value.

In [8]:
import altair as alt

chart = alt.Chart(data.reset_index().melt('day', var_name='airport',value_name='count').merge(airports.data[['icao','municipality']],
                                                                                                left_on='airport',right_on='icao',
                                                                                                how='left',)[['day','airport','count','municipality']].rename(columns=dict(municipality='city')))


### Trend line for airport usage (2019-2020); random selection of 5 dense hubs:

In [9]:
def display_chart(source, subset, subset_name):
    
    chart = source.transform_filter(alt.FieldOneOfPredicate(field='airport',oneOf=subset))
    highlight = alt.selection(type='single',nearest=True,on='mouseover',fields=['airport'])
    
    points = (chart.mark_point().encode(x='day',y=alt.Y('count',title='# of departing flights'),
                                      color=alt.Color('airport',legend=alt.Legend(title=subset_name)),
                                      tooltip=['day','airport','city','count'],
                                      opacity=alt.value(0.5),).add_selection(highlight))
    lines = (chart.mark_line().encode(x='day',y='count',color='airport',size=alt.condition(~highlight,
                                                                                          alt.value(1),alt.value(3)),).transform_loess('day','count',groupby=['airport'],bandwidth=0.2))
    return lines + points

result = alt.vconcat(*[display_chart(chart,airport_,subset_name).properties(width=750,height=400) for subset_name, airport_ in zip(['US hubs'],hubs,)]).resolve_scale(color='independent')

result

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


alt.VConcatChart(...)

## Route heatmaps - Prep and visualization embeds:
Routes generated by specific dates over an airspace defined by bounding box coordinates.

### Chicagoland region:

In [12]:
from traffic.data import opensky

In [13]:
# this requires access to data; request sent - awaiting approval ## APPROVAL RECEIVED
pre_day = "2020-02-25"
post_day = "2020-04-07"

In [14]:
bounds = [-88, 41., -87, 42.]

pre_data = opensky.history(start=pre_day, bounds=bounds, other_params=' and onground=false ')
post_data = opensky.history(start=post_day, bounds=bounds, other_params=' and onground=false ')

  0%|          | 0/24 [00:00<?, ?it/s]

  0%|          | 0/24 [00:00<?, ?it/s]

In [15]:
before_covid = (pre_data.clean_invalid().assign_id().filter().resample('10s').eval(desc='', max_workers=6))
after_covid = (post_data.clean_invalid().assign_id().filter().resample('10s').eval(desc='', max_workers=6))

  0%|          | 0/3674 [00:00<?, ?it/s]

  0%|          | 0/1542 [00:00<?, ?it/s]

In [ ]:
# Trend line df - write out later
vis1_data = data.reset_index().melt('day', var_name='airport',value_name='count').merge(airports.data[['icao','municipality']],
                                                                                                left_on='airport',right_on='icao',
                                                                                                how='left',)[['day','airport','count','municipality']].rename(columns=dict(municipality='city'))
vis1_data

In [16]:
for i, dat in enumerate([before_covid]):
    density= dat.query('altitude > 10000').agg_latlon(resolution=dict(latitude=90, longitude=90),
                                                         flight_id='nunique').query(f"flight_id > 10")

D:\Anaconda\envs\traffic\lib\site-packages\traffic\core\traffic.py:839: DeprecationWarning: agg_latlon will disappear in future versions. Use agg_xy instead
  warnings.warn(


In [17]:
for i, dat in enumerate([after_covid]):
    density_= dat.query('altitude > 10000').agg_latlon(resolution=dict(latitude=90, longitude=90),
                                                         flight_id='nunique').query(f"flight_id > 10")

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
D:\Anaconda\envs\traffic\lib\site-packages\traffic\core\traffic.py:839: DeprecationWarning: agg_latlon will disappear in future versions. Use agg_xy instead
  warnings.warn(


In [18]:
type(density)

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


pandas.core.frame.DataFrame

In [55]:
# Heatmap df - Chicagoland write out
# density.to_csv('densityChicPC19.csv',index=True)

In [56]:
# Heatmap df - Chicagoland post covid
# density_.to_csv('densityChicPoC19.csv', index=True)

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


### Chicagoland densities Pre and Post COVID:
Dates: February 25th 2020 - April 7th 2020

In [19]:
%%HTML
<div class='tableauPlaceholder' id='viz1623446059801' style='position: relative'><noscript><a href='#'><img alt='Route Density (Chicagoland) Pre COVID19 ' src='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;NB&#47;NBR2S8HH9&#47;1_rss.png' style='border: none' /></a></noscript><object class='tableauViz'  style='display:none;'><param name='host_url' value='https%3A%2F%2Fpublic.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='path' value='shared&#47;NBR2S8HH9' /> <param name='toolbar' value='yes' /><param name='static_image' value='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;NB&#47;NBR2S8HH9&#47;1.png' /> <param name='animate_transition' value='yes' /><param name='display_static_image' value='yes' /><param name='display_spinner' value='yes' /><param name='display_overlay' value='yes' /><param name='display_count' value='yes' /><param name='language' value='en-GB' /></object></div>                <script type='text/javascript'>                    var divElement = document.getElementById('viz1623446059801');                    var vizElement = divElement.getElementsByTagName('object')[0];                    vizElement.style.width='100%';vizElement.style.height=(divElement.offsetWidth*0.75)+'px';                    var scriptElement = document.createElement('script');                    scriptElement.src = 'https://public.tableau.com/javascripts/api/viz_v1.js';                    vizElement.parentNode.insertBefore(scriptElement, vizElement);                </script>

In [20]:
%%HTML
<div class='tableauPlaceholder' id='viz1623446150947' style='position: relative'><noscript><a href='#'><img alt='Route Density (Chicagoland) Post COVID19 ' src='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-ChicagolandPost-COVID&#47;RouteDensityChicagolandPostCOVID19&#47;1_rss.png' style='border: none' /></a></noscript><object class='tableauViz'  style='display:none;'><param name='host_url' value='https%3A%2F%2Fpublic.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='' /><param name='name' value='FlightRouteDensity-ChicagolandPost-COVID&#47;RouteDensityChicagolandPostCOVID19' /><param name='tabs' value='no' /><param name='toolbar' value='yes' /><param name='static_image' value='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-ChicagolandPost-COVID&#47;RouteDensityChicagolandPostCOVID19&#47;1.png' /> <param name='animate_transition' value='yes' /><param name='display_static_image' value='yes' /><param name='display_spinner' value='yes' /><param name='display_overlay' value='yes' /><param name='display_count' value='yes' /><param name='language' value='en-GB' /></object></div>                <script type='text/javascript'>                    var divElement = document.getElementById('viz1623446150947');                    var vizElement = divElement.getElementsByTagName('object')[0];                    vizElement.style.width='100%';vizElement.style.height=(divElement.offsetWidth*0.75)+'px';                    var scriptElement = document.createElement('script');                    scriptElement.src = 'https://public.tableau.com/javascripts/api/viz_v1.js';                    vizElement.parentNode.insertBefore(scriptElement, vizElement);                </script>

Generated with bounding box coordinates: Area covers KORD, KMDW and KGYY (Chicago, Midway, Naperville, Bolingbrook, Gary and Schaumburg. 

### Los Angeles region:

In [21]:
LA_bounds = [-118, 33., -116, 34.]

pre_data_LA = opensky.history(start=pre_day, bounds=LA_bounds, other_params=' and onground=false ')
post_data_LA = opensky.history(start=post_day, bounds=LA_bounds, other_params=' and onground=false ')

  0%|          | 0/24 [00:00<?, ?it/s]

  0%|          | 0/24 [00:00<?, ?it/s]

In [23]:
before_covid_LA = (pre_data_LA.clean_invalid().assign_id().filter().resample('10s').eval(desc='', max_workers=6))
after_covid_LA = (post_data_LA.clean_invalid().assign_id().filter().resample('10s').eval(desc='', max_workers=6))

  0%|          | 0/3673 [00:00<?, ?it/s]

  0%|          | 0/966 [00:00<?, ?it/s]

In [24]:
for i, dat in enumerate([before_covid_LA]):
    densityLA = dat.query('altitude > 10000').agg_latlon(resolution=dict(latitude=90, longitude=90),
                                                         flight_id='nunique').query(f"flight_id > 10")

D:\Anaconda\envs\traffic\lib\site-packages\traffic\core\traffic.py:839: DeprecationWarning: agg_latlon will disappear in future versions. Use agg_xy instead
  warnings.warn(


In [25]:
for i, dat in enumerate([after_covid_LA]):
    densityLA_= dat.query('altitude > 10000').agg_latlon(resolution=dict(latitude=90, longitude=90),
                                                         flight_id='nunique').query(f"flight_id > 10")

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
D:\Anaconda\envs\traffic\lib\site-packages\traffic\core\traffic.py:839: DeprecationWarning: agg_latlon will disappear in future versions. Use agg_xy instead
  warnings.warn(


In [16]:
# LA region heatmap df - Pre covid write out
# densityLA.to_csv('densityLAPC19.csv',index=True)

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [17]:
# LA region heatmap df - post covid write out
# densityLA_.to_csv('densityLAPoC19.csv', index=True)

### Los Angeles densities Pre and Post COVID:
Dates: February 25th 2020 - April 7th 2020

In [26]:
%%HTML
<div class='tableauPlaceholder' id='viz1623452290236' style='position: relative'><noscript><a href='#'><img alt='Route Density (Los Angeles) Pre COVID19 ' src='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;YT&#47;YT4W5Z2Z8&#47;1_rss.png' style='border: none' /></a></noscript><object class='tableauViz'  style='display:none;'><param name='host_url' value='https%3A%2F%2Fpublic.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='path' value='shared&#47;YT4W5Z2Z8' /> <param name='toolbar' value='yes' /><param name='static_image' value='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;YT&#47;YT4W5Z2Z8&#47;1.png' /> <param name='animate_transition' value='yes' /><param name='display_static_image' value='yes' /><param name='display_spinner' value='yes' /><param name='display_overlay' value='yes' /><param name='display_count' value='yes' /><param name='language' value='en-GB' /></object></div>                <script type='text/javascript'>                    var divElement = document.getElementById('viz1623452290236');                    var vizElement = divElement.getElementsByTagName('object')[0];                    vizElement.style.width='100%';vizElement.style.height=(divElement.offsetWidth*0.75)+'px';                    var scriptElement = document.createElement('script');                    scriptElement.src = 'https://public.tableau.com/javascripts/api/viz_v1.js';                    vizElement.parentNode.insertBefore(scriptElement, vizElement);                </script>

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [27]:
%%HTML
<div class='tableauPlaceholder' id='viz1623452346009' style='position: relative'><noscript><a href='#'><img alt='Route Density (Los Angeles) Post COVID19 ' src='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-LosAngelesPost-COVID&#47;RouteDensityLosAngelesPostCOVID19&#47;1_rss.png' style='border: none' /></a></noscript><object class='tableauViz'  style='display:none;'><param name='host_url' value='https%3A%2F%2Fpublic.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='' /><param name='name' value='FlightRouteDensity-LosAngelesPost-COVID&#47;RouteDensityLosAngelesPostCOVID19' /><param name='tabs' value='no' /><param name='toolbar' value='yes' /><param name='static_image' value='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-LosAngelesPost-COVID&#47;RouteDensityLosAngelesPostCOVID19&#47;1.png' /> <param name='animate_transition' value='yes' /><param name='display_static_image' value='yes' /><param name='display_spinner' value='yes' /><param name='display_overlay' value='yes' /><param name='display_count' value='yes' /><param name='language' value='en-GB' /></object></div>                <script type='text/javascript'>                    var divElement = document.getElementById('viz1623452346009');                    var vizElement = divElement.getElementsByTagName('object')[0];                    vizElement.style.width='100%';vizElement.style.height=(divElement.offsetWidth*0.75)+'px';                    var scriptElement = document.createElement('script');                    scriptElement.src = 'https://public.tableau.com/javascripts/api/viz_v1.js';                    vizElement.parentNode.insertBefore(scriptElement, vizElement);                </script>

Los Angeles area - primary airport in the area is KLAX; two secondary airports RIV and CNO are also in the area. Airspace is one of the busiest in the world; California mandates were one of the first states to be implemented and as is with any California law - they came down hard on lockdowns. 
Note - routes over designated parks are likely routes taken by firefighting aircraft to dump fire retardant, foam, and water over forest fires.

### New York region:

In [28]:
NY_bounds = [-74, 40., -71, 41.]

pre_data_NY = opensky.history(start=pre_day, bounds=NY_bounds, other_params=' and onground=false ')
post_data_NY = opensky.history(start=post_day, bounds=NY_bounds, other_params=' and onground=false ')

  0%|          | 0/24 [00:00<?, ?it/s]

  0%|          | 0/24 [00:00<?, ?it/s]

In [29]:
before_covid_NY = (pre_data_NY.clean_invalid().assign_id().filter().resample('10s').eval(desc='', max_workers=6))
after_covid_NY = (post_data_NY.clean_invalid().assign_id().filter().resample('10s').eval(desc='', max_workers=6))

  0%|          | 0/3743 [00:00<?, ?it/s]

  0%|          | 0/777 [00:00<?, ?it/s]

In [30]:
for i, dat in enumerate([before_covid_NY]):
    densityNY = dat.query('altitude > 10000').agg_latlon(resolution=dict(latitude=90, longitude=90),
                                                         flight_id='nunique').query(f"flight_id > 10")

D:\Anaconda\envs\traffic\lib\site-packages\traffic\core\traffic.py:839: DeprecationWarning: agg_latlon will disappear in future versions. Use agg_xy instead
  warnings.warn(


In [31]:
for i, dat in enumerate([after_covid_NY]):
    densityNY_= dat.query('altitude > 10000').agg_latlon(resolution=dict(latitude=90, longitude=90),
                                                         flight_id='nunique').query(f"flight_id > 10")

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
D:\Anaconda\envs\traffic\lib\site-packages\traffic\core\traffic.py:839: DeprecationWarning: agg_latlon will disappear in future versions. Use agg_xy instead
  warnings.warn(


In [28]:
# NY Pre covid df write out
# densityNY.to_csv('densityNYPC19.csv',index=True)

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [30]:
# NY Post covid df write out
# densityNY_.to_csv('densityNYPoC19.csv', index=True)

### New York Pre and Post COVID:
Dates: February 25th 2020 - April 7th 2020

In [32]:
%%HTML
<div class='tableauPlaceholder' id='viz1623456651266' style='position: relative'><noscript><a href='#'><img alt='Route Density (New York) Pre-COVID19 ' src='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-NewYorkPre-COVID&#47;RouteDensityNewYorkPre-COVID19&#47;1_rss.png' style='border: none' /></a></noscript><object class='tableauViz'  style='display:none;'><param name='host_url' value='https%3A%2F%2Fpublic.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='' /><param name='name' value='FlightRouteDensity-NewYorkPre-COVID&#47;RouteDensityNewYorkPre-COVID19' /><param name='tabs' value='no' /><param name='toolbar' value='yes' /><param name='static_image' value='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-NewYorkPre-COVID&#47;RouteDensityNewYorkPre-COVID19&#47;1.png' /> <param name='animate_transition' value='yes' /><param name='display_static_image' value='yes' /><param name='display_spinner' value='yes' /><param name='display_overlay' value='yes' /><param name='display_count' value='yes' /><param name='language' value='en-GB' /></object></div>                <script type='text/javascript'>                    var divElement = document.getElementById('viz1623456651266');                    var vizElement = divElement.getElementsByTagName('object')[0];                    vizElement.style.width='100%';vizElement.style.height=(divElement.offsetWidth*0.75)+'px';                    var scriptElement = document.createElement('script');                    scriptElement.src = 'https://public.tableau.com/javascripts/api/viz_v1.js';                    vizElement.parentNode.insertBefore(scriptElement, vizElement);                </script>

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [33]:
%%HTML
<div class='tableauPlaceholder' id='viz1623456711900' style='position: relative'><noscript><a href='#'><img alt='Route Density (New York) Post-COVID ' src='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-NewYorkPost-COVID&#47;Sheet6&#47;1_rss.png' style='border: none' /></a></noscript><object class='tableauViz'  style='display:none;'><param name='host_url' value='https%3A%2F%2Fpublic.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='' /><param name='name' value='FlightRouteDensity-NewYorkPost-COVID&#47;Sheet6' /><param name='tabs' value='no' /><param name='toolbar' value='yes' /><param name='static_image' value='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Fl&#47;FlightRouteDensity-NewYorkPost-COVID&#47;Sheet6&#47;1.png' /> <param name='animate_transition' value='yes' /><param name='display_static_image' value='yes' /><param name='display_spinner' value='yes' /><param name='display_overlay' value='yes' /><param name='display_count' value='yes' /><param name='language' value='en-GB' /></object></div>                <script type='text/javascript'>                    var divElement = document.getElementById('viz1623456711900');                    var vizElement = divElement.getElementsByTagName('object')[0];                    vizElement.style.width='100%';vizElement.style.height=(divElement.offsetWidth*0.75)+'px';                    var scriptElement = document.createElement('script');                    scriptElement.src = 'https://public.tableau.com/javascripts/api/viz_v1.js';                    vizElement.parentNode.insertBefore(scriptElement, vizElement);                </script>

There are 3 primary hubs in the New York region - KJFK, KLGA, and KEWR. By sheer number, this region is by far the busiest in the entirety of the United States. There is an additional airport that supplement the 3 large hubs: KSWF - which plays a similar role as KGYY to KORD and KMDW. New York was one of the cities that was hit the hardest by the pandemic both socially and economically. The low densities mirrors billions of dollars worth of revenue lost to the Aviation industry - not withstanding the loss of jobs resulting from it. 

### Chicagoland analysis:

In [52]:
density.shape

(1580, 1)

In [51]:
density_.shape

(335, 1)

There are 1580 logged flights over the Chicagoland area above an altitude of 10000 feet AGL on February 25th 2020, compared to the 335 logged flights on April 7th, 2020, over the airspace in the specified bounding box. There was a 78.8% reduction in the density of the airspace above the region. 
- Percentage aggregation used: (max - min) / max * 100

### Los Angeles region analysis:

In [49]:
densityLA.shape

(1868, 1)

In [50]:
densityLA_.shape

(513, 1)

There were about 1868 flights logged over the airspace coordinates that define the LA region on February 25th 2020; this value was reduced to just 513 flights on April 7th 2020, once lockdowns and travels bans went into effect, possibly even before that. There was a reduction of flight density by 72.5% in this region. 

### NY region analysis:

In [48]:
densityNY.shape

(2032, 1)

In [47]:
densityNY_.shape

D:\Anaconda\envs\traffic\lib\site-packages\ipykernel\ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(54, 1)

The New York area is home to 3 of the biggest airport hubs which constitute the busiest airspace for both commercial and cargo flight operations. There were 2032 logged flights in the airspace coordinates on February 25th 2020, which was reduced to just 54 on April 7th 2020. That is a 97.3% reduction in airspace density over the LA region, owing mostly to travel bans and strict lockdown measures. This region was by far the worst affected by revenue loss to the Aviation Industry, simply on grounds of the sheer number of flights that were grounded or cancelled.

## Analysis Conclusion:
`` Fill in once mandates are analyzed ``